src/context.rs

```diff
@@ -158,6 +162,57 @@ impl Context {
         self.blobdir.as_path()
     }
 
+    pub fn copy_to_blobdir(&self, orig_filename: impl AsRef<str>) -> Result<String> {
+        // return a $BLOBDIR/<filename> with the content of orig_filename
+        // copied into it. The <filename> will be safely derived from
+        // orig_filename, and will not clash with existing filenames.
+        let dest = self.new_blob_file(&orig_filename, b"")?;
+        if dc_copy_file(
+            &self,
+            PathBuf::from(orig_filename.as_ref()),
+            PathBuf::from(&dest),
+        ) {
+            Ok(dest)
+        } else {
+            bail!("could not copy {} to {}", orig_filename.as_ref(), dest);
+        }
+    }
+
+    pub fn new_blob_file(&self, orig_filename: impl AsRef<str>, data: &[u8]) -> Result<String> {
+        // return a $BLOBDIR/<FILENAME> string which corresponds to the
+        // respective file in the blobdir, and which contains the data.
+        // FILENAME is computed by looking and possibly mangling the
+        // basename of orig_filename. The resulting filenames are meant
+        // to be human-readable.
+        let (stem, ext) = dc_derive_safe_stem_ext(orig_filename.as_ref());
```

Comment:
You wan to use `Path::file_stem()` rather than write a custom version of this.  Likewise `Path::extension()` gives you the extension.

Reply:
i spent quite some time trying to use path but it's making things more complicated.  Also the "$BLOBDIR/basename.ext" things are not really paths, they can be made into path but are not proper paths themselves. I'd rather like to keep it as is.